### PR TITLE
Fix relative paths in wasm backend source maps. Fixes #9837

### LINKED ISSUE #9837: Invalid URLs (absolute paths) in source maps in some cases
## Summary

Source maps are supposed to contain only valid URLs in the `sources` array (either relative, or absolute with scheme). Emscripten sometimes emits absolute paths into that array instead, making them impossible to load by a browser.

## Examples

For a file `C:\Users\rreverser\Projects\temp-wasm\temp.c` with arbitary content in a folder `temp-wasm`, I'm trying different formats for the filename.

Note that this is not a Windows-specific issue, and you'll get same results with Unix filepaths.

1. Running `emcc ./temp.c -g4 -o temp.wasm` (relative filename in the same folder):

```js
{"version":3,"sources":["./temp.c"],"names":[],...
```

2. Running `emcc temp.c -g4 -o temp.wasm` (same relative filename, but without explicit `./`):

```js
{"version":3,"sources":["C:/Users/rreverser/Projects/temp-wasm/temp.c"],"names":[],...
```

3. Running `emcc ../temp-wasm/temp.c` (relative filename with a folder):

```js
{"version":3,"sources":["../temp-wasm/temp.c"],"names":[],
```

4. Running `emcc C:\Users\rreverser\Projects\temp-wasm\temp.c -g4 -o temp.wasm` (absolute filename):

```js
{"version":3,"sources":["C:/Users/rreverser/Projects/temp-wasm/temp.c"],"names":[],...
```

## Results

When a filename is relative but without a leading `./` or is specified as absolute, Emscripten outputs an absolute filename in `sources`, which is not a valid URL and can't be handled by DevTools. It could be valid if prefixed with `file:///` specifier, but that's not an ideal solution.

The only case where Emscripten `sources` ends up being valid URL is when a filename is specified as explicitly relative (cases 1 and 3).

Then, Emscripten outputs a relative URL as well, which can be handled well by DevTools both in case of loading from filesystem as well as a static HTTP server.

More so, even though case 3 outputs a relative filename, it's not ideal.

## Suggested solution

Emscripten should always compute relative URL out of the source map filename and the actual sources, and emit that in `sources` array.

### PR DESCRIPTION
**Pull request by kripken**

### CODE REVIEW

File: tests/test_other.py [hunk 1]
```diff
@@ -8916,6 +8916,21 @@ def test_wasm_sourcemap_dead(self):
...
     self.assertRegexpMatches(output, r'"mappings":\s*"[A-Za-z0-9+/]+,[A-Za-z0-9+/]+"')
 
+  def test_wasm_sourcemap_relative_paths(self):
+    filename = 'a.cpp'
+    shutil.copyfile(path_from_root('tests', 'hello_123.c'), 'a.cpp')
+    filenames = [
+      filename,
+      os.path.abspath(filename)
+    ]
+    if not WINDOWS:
```
Comment (RReverser) [inline tests/test_other.py]: Why is this platform-specific?

File: tools/wasm-sourcemap.py [hunk 2]
```diff
@@ -263,7 +263,7 @@ def build_sourcemap(entries, code_section_offset, prefixes, collect_sources):
     if column == 0:
       column = 1
     address = entry['address'] + code_section_offset
-    file_name = entry['file']
+    file_name = os.path.relpath(entry['file'])
```
Comment (RReverser) [inline tools/wasm-sourcemap.py]: I'm not sure this is correct: it uses a path relative to the current directory, but browsers / spec expect `sources` to be relative to the source map file.

For example, consider case like following:
```bash
$ emcc ./temp.c -g4 -o dist/temp.wasm
```

Currently this produces a source map `dist/temp.wasm.map` with following contents:
```js
{"version":3,"sources":["./temp.c"],...
```

But this is incorrect, because `temp.c` is actually in a folder up and needs to be referenced as `../temp.c`.

File: tests/test_other.py (hunk 1 above)
Comment (kripken) [reply tests/test_other.py]: I don't think `./filename` is valid on windows?

File: tools/wasm-sourcemap.py (hunk 2 above)
Comment (kripken) [reply tools/wasm-sourcemap.py]: In this case wouldn't the user need to set up `--source-map-base` to fix that? But I've never really understood any of this stuff, hopefully you do...

File: tests/test_other.py (hunk 1 above)
Comment (RReverser) [reply tests/test_other.py]: It is! Windows happily accepts forward-slashes (we discussed this when I made the change to unconditionally replace `\` with `/` too).

File: tools/wasm-sourcemap.py (hunk 2 above)
Comment (RReverser) [reply tools/wasm-sourcemap.py]: So `--source-map-base` helps browser to find a source map from WebAssembly module.

But once browser found that source map, it now needs to find source files. There are three potential ways for it to do that:
1. `sources` are specified as absolute URLs (not great for local files, but can work well enough for public libs).
2. `sources` are specified as relative URLs to the source map itself.
3. Source map contains `sourceRoot`, and then all `sources` are relative to that instead (could be used in our case as an alternative, but it's a bit more complicated IMO).

Comment (RReverser) [reply tools/wasm-sourcemap.py]: Actually, maybe you're right, and in this case `--source-map-base` might be enough, at least if `.wasm.map` is always emitted alongside `.wasm`. Hmm...

Comment (RReverser) [reply tools/wasm-sourcemap.py]: No, wait, scratch that, because relative URL still has to be relative to the source map itself. So the comment before still stands :)

Comment (RReverser) [reply tools/wasm-sourcemap.py]: What browser is basically doing:
1) Extract `sourceMappingURL` from the WebAssembly module (this one is `{--source-map-base}/out.wasm.map`).
2) Load that source map.
3) If that source map contains `sourceRoot`, prepend it to each item in `sources`.
4) Resolve each item in `sources` relative to `sourceMappingURL`.

So we need each item in `sources` to be relative to `out.wasm.map`, no matter where it is.

File: tests/test_other.py (hunk 1 above)
Comment (kripken) [reply tests/test_other.py]: Oh, even with the dot for the current dir?

File: tools/wasm-sourcemap.py (hunk 2 above)
Comment (kripken) [reply tools/wasm-sourcemap.py]: Thanks! I think I see a little better. I'm still confused about how that `sourceMappingURL` option is used in two places for seemingly different purposes, though?

File: tests/test_other.py (hunk 1 above)
Comment (RReverser) [reply tests/test_other.py]: Yeah, sure, relative paths are pretty much the same everywhere (that is, `.` and `..` work in the same way).

File: tools/wasm-sourcemap.py (hunk 2 above)
Comment (RReverser) [reply tools/wasm-sourcemap.py]: In both places it's used as "URL of the .wasm.map". First, to find the map, and later to resolve other files it references, but semantic meaning is the same.

It's similar to how most module systems work, too: if A imports B, and B imports C, then first B is found from A, and then reference to C is resolve relatively to B.

File: tools/shared.py [hunk 3]
```diff
@@ -2871,10 +2871,14 @@ def path_to_system_js_libraries(library_name):
 
   @staticmethod
   def emit_wasm_source_map(wasm_file, map_file):
+    # source file paths must be relative to the location of the map (which is
+    # emitted alongside the wasm)
+    base_path = os.path.dirname(os.path.abspath(Settings.WASM_BINARY_FILE))
```
Comment (RReverser) [inline tools/shared.py]: It probably doesn't matter since they're emitted alongside, but can't we use `map_file` here?

Comment (kripken) [reply tools/shared.py]: sadly no; we run these commands on temp files. the new code here plugs through the final path of the output map file.

**Comment by kripken** [comment]
Those last commits should fix the test failures, at least the ones I ran locally.

I had to change some tests, which I think was correct, but please take a look at those as I'm not sure.

File: tools/shared.py (hunk 3 above)
Comment (RReverser) [reply tools/shared.py]: Makes sense, thanks.

**Comment by kripken** [comment]
Ok, tests look good!

File: tests/test_core.py [hunk 4]
```diff
@@ -7179,6 +7179,12 @@ def encode_utf8(data):
       else:
         return data
 
+    def source_map_file_loc(name):
+      if shared.Settings.WASM_BACKEND:
+        return name
+      # in fastcomp, we have the absolute path, which is not good
```
Comment (RReverser) [inline tests/test_core.py]: I don't really understand this comment - can you please clarify why/how fastcomp is special?

File: tests/test_other.py [hunk 5]
```diff
@@ -8910,12 +8911,35 @@ def test_wasm_sourcemap_dead(self):
...
                     '-o', 'a.out.wasm.map',
-                    path_from_root('tests', 'other', 'wasm_sourcemap_dead', 't.wasm')]
+                    path_from_root('tests', 'other', 'wasm_sourcemap_dead', 't.wasm'),
+                    '--basepath=' + os.getcwd()]
     run_process(wasm_map_cmd, stdout=PIPE, stderr=PIPE)
     output = open('a.out.wasm.map').read()
     # has only two entries
     self.assertRegexpMatches(output, r'"mappings":\s*"[A-Za-z0-9+/]+,[A-Za-z0-9+/]+"')
 
+  @no_fastcomp()
```
Comment (RReverser) [inline tests/test_other.py]: Same here re:fastcomp.

File: tests/test_other.py [hunk 6]
```diff
@@ -8910,12 +8911,35 @@ def test_wasm_sourcemap_dead(self):
...
+  def test_wasm_sourcemap_relative_paths(self):
+    def test(infile, source_map_added_dir=''):
+      expected_source_map_path = os.path.join(source_map_added_dir, 'a.cpp')
+      print(infile, expected_source_map_path)
```
Comment (RReverser) [inline tests/test_other.py]: Are these leftover prints or necessary in tests?

File: tools/wasm-sourcemap.py [hunk 7]
```diff
@@ -264,6 +264,14 @@ def build_sourcemap(entries, code_section_offset, prefixes, collect_sources):
       column = 1
     address = entry['address'] + code_section_offset
     file_name = entry['file']
+    # prefer to emit a relative path to the base path, which is where the
```
Comment (RReverser) [inline tools/wasm-sourcemap.py]: I'm not sure if we should do anything special here... since we're talking about edge cases anyway, consider one where a static server that is hosting `/` directory under, say, `http://somedomain/static/`.

Then, let's say Wasm is under `/dist/temp.wasm` on the disk (corresponding to URL: `http://somedomain/static/dist/temp.wasm` on server), and source is under `/src/temp.c` (URL: `http://somedomain/static/src/temp.c`).

If you use relative paths, then source would be encoded as `../src/temp.c` and resolve correctly in URL context, but if you keep it as-is, then it would be encoded as `/src/temp.c` and resolve to `http://somedomain/src/temp.c` in browser, which is invalid.

---

TL;DR - let's just do the simple thing and use relative paths unconditionally, which would happen to also cover edge cases more correctly.

File: tests/test_core.py (hunk 4 above)
Comment (kripken) [reply tests/test_core.py]: Oh, fastcomp just uses a different code path for source map generation. We are deprecating and planning to remove fastcomp, so I don't think it's worth doing double work to fix that code path too.

File: tests/test_other.py (hunk 6 above)
Comment (kripken) [reply tests/test_other.py]: They aren't necessary, but they make following the output easier when a test fails on CI, so a bunch of tests have them.

File: tools/wasm-sourcemap.py (hunk 7 above)
Comment (kripken) [reply tools/wasm-sourcemap.py]: That would break the current `other.test_wasm_sourcemap` test, it gets this as the source file location:
```
"wasm-src://../../emscripten/tests/other/wasm_sourcemap/no_main.c"
```
Is that ok?

Comment (kripken) [reply tools/wasm-sourcemap.py]: (if that's not ok, what would the proper output there be, do you think?)

Comment (RReverser) [reply tools/wasm-sourcemap.py]: I see, interesting. Where is `wasm-src://` added? We probably should skip resolving relative paths altogether when that option is enabled.

File: tests/test_core.py (hunk 4 above)
Comment (RReverser) [reply tests/test_core.py]: Ah ok. But I think unconditionally returning `name` should still just work in both modes? (`os.path.relpath` seems to perform `abspath` internally anyway, so it doesn't require an absolute path as an input)

Comment (RReverser) [reply tests/test_core.py]: Ah, this is test, not implementation, so it's output paths that are different. Seems fine then :)

File: tools/wasm-sourcemap.py (hunk 7 above)
Comment (kripken) [reply tools/wasm-sourcemap.py]: It's added in the test, using the `--prefix` flag to the source map tool (which I don't understand well).

Maybe we can remove that option? If you and @yurydelendik agree that should be fine as I don't see other users in the codebase.

Comment (yurydelendik) [reply tools/wasm-sourcemap.py]: The DWARF information must contain absolute path to source file in .debug_line and .debug_info. If for some reason we don't have that, it needs to be fixed. The `--prefix` was designed to perform the basepath logic, e.g. "http://somedomain/static/dist/temp.wasm + http://somedomain/src/temp.c" with use of `--prefix /Users/me/example/=../src/` shall produce the desired result. I wonder if wasm-sourcemap.py not reading/building entire DWARF file path.

Comment (kripken) [reply tools/wasm-sourcemap.py]: @yurydelendik isn't adding a prefix what `--source-map-base` is for? (which does not use `--prefix` under the hood) Or is that for something else?

Comment (yurydelendik) [reply tools/wasm-sourcemap.py]: `--source-map-base` is instruction for tools to add proper map location in sourceMappingURL, this is optional. (As defined in https://github.com/WebAssembly/design/pull/1051) Devtools use `sourceMappingURL` as base URL for all source names.

`--prefix`, on server/build side, helps to strip useless for devtools absolute part. Notice that sources paths needs to point to the valid internet resource only `sourcesContent` is not present.

Comment (kripken) [reply tools/wasm-sourcemap.py]: It looks like nothing uses `--prefix`, though - I think we are avoiding the problem you describe with the other tools here already? Or are you aware of out-of-tree users of wasm-sourcemap.py? If not, would you object to me removing that flag?

Comment (yurydelendik) [reply tools/wasm-sourcemap.py]: `--prefix` is designed to match gcc's `-fdebug-prefix-map=` . I don't have objection, except convenience, since clang or rustc can do it. Will `emcc` passthru this option?

> are you aware of out-of-tree users of wasm-sourcemap.py

wasm-sourcemap.py is designed and is used in only for emscripten (and my examples).

Comment (kripken) [reply tools/wasm-sourcemap.py]: Interesting, I wasn't aware of `-fdebug-prefix-map`. emcc does not currently pass that through. Perhaps it should?

Also, getting back to the more immediate issue here: It seems like the fundamental issue is that when supplying `--prefix`, we should not automatically relativize a path. That is what this PR does currently, but @RReverser wrote this comment suggesting we always relativize. To me it sounds like the PR should stand as it is - @RReverser what do you think?

Comment (RReverser) [reply tools/wasm-sourcemap.py]: 1. No, if we keep it, I also suggested that we shouldn't relativize when `--prefix` is passed:
> We probably should skip resolving relative paths altogether when that option is enabled.
2. However, it's not what the PR does at the moment - right now it checks whether paths don't have a shared prefix. This check is incorrect and breaks cases like described in comment (1) of this thread. Instead, it should skip relativising specifically when `--prefix` is set.
3. Finally, maybe I misunderstood, but from the exchange above it sounds like we can remove `--prefix` instead?

Comment (yurydelendik) [reply tools/wasm-sourcemap.py]: > emcc does not currently pass that through. Perhaps it should?

Yes, it should. After removal `--prefix` it will be the only way to properly re-point source to right location.

Here is a use case: there are several libraries that are checked out locally on a build machine. The desired results is that source with point to proper internet resources. The only way to do provide right source location is to use `-fdebug-prefix-map`, e.g. `-fdebug-prefix-map=/Users/build/lib1/=http://example.org/lib1/v1/ --fdebug-prefix-map=/Users/build/lib2/=http://example.org/lib2/v9/`.

Comment (RReverser) [reply tools/wasm-sourcemap.py]: @yurydelendik Doesn't `-fdebug-prefix-map` in Clang change DWARF paths in-place to the remapped locations? If it does, then what is the role of custom `--prefix` handling?

Comment (yurydelendik) [reply tools/wasm-sourcemap.py]: > what is the role of custom --prefix handling

To do it after the fact of creating a wasm binary. (`-fdebug-prefix-map` did not work in rustc at the moment of writing wasm-sourcemap.py)

Comment (RReverser) [reply tools/wasm-sourcemap.py]: Sure, but now that both compilers support native remapping, it seems that we can just skip it here?

Comment (yurydelendik) [reply tools/wasm-sourcemap.py]: the `--base_path` logic is a really narrow case of the `--prefix`. The former assumes that sources folders will have the same directory structure when published and calculates replacement part automatically, the latter provides more flexibility.  FWIW they can exist together (by adding `else:` for `prefixes.sources.resolve` line). Sorry, I cannot provide more definite "yes, let's remove it", since I accustomed to the `--prefix` way. :)

Comment (kripken) [reply tools/wasm-sourcemap.py]: Thanks @RReverser , you are right, I had that mixed up! Should be correct now - use the prefixes if provided, otherwise use a proper relative path.

**Comment by kripken** [comment]
Ah, some other tests now break. There is still the question of what to do with "bad" relative paths.

For example, if I am in `/foo/` and I compile a file not under that directory, `/bar/file.cpp`, then if we use a relative path as in this PR, we get `../bar/file.cpp`. Is that better than emitting `/bar/file.cpp`? (Or are both unuseful anyhow?)

**Comment by RReverser** [comment]
>  Is that better than emitting /bar/file.cpp? (Or are both unuseful anyhow?)

Yes - I described the difference for precisely same case in above comment thread https://github.com/emscripten-core/emscripten/pull/9882#discussion_r349583022.

TL;DR #2 - `/bar/file.cpp` would be still invalid in browser unless your `.wasm` + `.wasm.map` is hosted at the root of the website, while `../bar/file.cpp` would work in any nested folders too.

**Comment by RReverser** [comment]
On a second thought, I suppose these cases are actually rare, so maybe we can just print error in case of sources and output on different root folders and/or Windows disks? Sounds like this would solve our final problems.

**Comment by kripken** [comment]
Thanks @RReverser , I updated the lsan tests to accept those relative paths then.

I agree just showing an error might be another reasonable path here, but I'm not sure how best to explain to users what to do in the error message, seems nontrivial - I'd rather leave that to followup work if we see it's necessary in practice, with people hitting this issue.

**Comment by RReverser** [comment]
@kripken OK, makes sense to me.

**Comment by sbc100** [comment]
This breaks the windows builder:

https://logs.chromium.org/logs/emscripten-releases/buildbucket/cr-buildbucket.appspot.com/8895598112699504816/+/steps/Emscripten_testsuite__upstream__other_/0/stdout

**Comment by sbc100** [comment]
I think I'm going to revert this so we can get a green build over the weekend.
